### ISSUE TITLE
10$: Warn on mixing >> and & precedence

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#1056]
`a & 0x0f00 >> 16` is probably a bug. It is interpreted as `a & (0x0f00 >> 16)` but the programmer meant `(a & 0x0f00) >> 16`. We should warn in the following cases, where _C#_ is some number literal:

```
a & C1 >> C2
a & C1 << C2
```

**Comment by toastin0** [comment]
do you mind if I try this one out as a first contribution? i really like the project, and think that contributing might be a solid way to help reinforce knowledge of C++.

**Comment by strager** [comment]
> do you mind if I try this one out as a first contribution?

You may! Let me know if you have any questions or trouble. Here are our contribution docs: https://quick-lint-js.com/contribute/

**Comment by strager** [comment]
@toastin0 You asked in IRC for clarification.

Take `3 & 2 << 1` as an example. There are two interpretations:

1. `(3 & 2) << 1` == 4
2. `3 & (2 << 1)` == 0

JavaScript takes the second interpretation, thus `3 & 2 << 1` equals 0. (Other languages like C also have this interpretation.)

However, if the programmer wrote `x & 2 << 1` (where `x` is some expression and `2` and `1` are literals), then in my opinion the programmer probably wanted the first interpretation (`(x & 2) << 1`).

It is possible that the programmer did want `x & (2 << 1)`, but I doubt it. I wonder if there are counterexamples somewhere.

**Comment by toastin0** [comment]
thank you very much for the clarification! good news is, i've already written tests in line with what you said, so this doesn't change anything on my end, but i wanted to make doubly sure that what i wrote was correct. i'll get to work on implementation now!

**Comment by strager** [comment]
I searched this [JavaScript corpus](https://github.com/quick-lint/javascript-corpus) and found no instances of this pattern. (I did find `x>>C1&C2` though.) However, my search is probably imperfect. It'd be cool to run your patch on the corpus and see if it catches any false positives or false negatives. (I can do this.)
